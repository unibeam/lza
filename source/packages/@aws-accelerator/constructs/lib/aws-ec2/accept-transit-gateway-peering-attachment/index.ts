/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
/**
 * delete-default-vpc - lambda handler
 *
 * @param event
 * @returns
 */

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  const requesterAccountId = event.ResourceProperties['requesterAccountId'];
  const requesterRegion = event.ResourceProperties['requesterRegion'];
  const requesterTransitGatewayRouteTableId = event.ResourceProperties['requesterTransitGatewayRouteTableId'];
  const requesterTransitGatewayId = event.ResourceProperties['requesterTransitGatewayId'];

  const accepterAccountId = event.ResourceProperties['accepterAccountId'];
  const accepterRegion = event.ResourceProperties['accepterRegion'];
  const accepterTransitGatewayId = event.ResourceProperties['accepterTransitGatewayId'];
  const accepterTransitGatewayRouteTableId = event.ResourceProperties['accepterTransitGatewayRouteTableId'];

  const autoAccept: boolean = event.ResourceProperties['autoAccept'] === 'true';
  const peeringTags: { Key: string; Value: string }[] | undefined = event.ResourceProperties['peeringTags'];

  const requesterTransitGatewayAttachmentId = event.ResourceProperties['requesterTransitGatewayAttachmentId'];

  const solutionId = process.env['SOLUTION_ID'];

  const requesterEc2Client = new AWS.EC2({ region: requesterRegion });

  const accepterEc2Client = await getAccepterEc2Client(
    accepterRegion,
    event.ResourceProperties['accepterRoleArn'],
    solutionId,
  );

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const accepterTransitGatewayAttachmentId = await getAccepterTransitGatewayAttachmentID(
        accepterEc2Client,
        accepterTransitGatewayId,
        accepterAccountId,
        requesterTransitGatewayId,
        requesterTransitGatewayAttachmentId,
      );

      if (autoAccept) {
        console.log(`Starting - Transit Gateway Peering attachment acceptance`);

        let attachmentStatus = await getAttachmentState(requesterEc2Client, requesterTransitGatewayAttachmentId);

        await checkAttachmentInvalidStateForAcceptance(requesterEc2Client, requesterTransitGatewayAttachmentId);

        while (attachmentStatus !== 'pendingAcceptance' && attachmentStatus !== 'available') {
          await delay(1000);
          console.log(
            `Waiting for transit gateway attachment id ${requesterTransitGatewayAttachmentId} to be in pendingAcceptance or available state in requester account ${requesterAccountId}, current state is : ${attachmentStatus}`,
          );
          attachmentStatus = await getAttachmentState(requesterEc2Client, requesterTransitGatewayAttachmentId);
          await checkAttachmentInvalidStateForAcceptance(requesterEc2Client, requesterTransitGatewayAttachmentId);
        }

        console.log(
          `Attachment state is : ${attachmentStatus} in requester account ${requesterAccountId} for transit gateway attachment id ${requesterTransitGatewayAttachmentId}, proceeding with acceptTransitGatewayPeeringAttachment operation in accepter account ${accepterAccountId}`,
        );

        if (!accepterTransitGatewayAttachmentId) {
          throw new Error(
            `Transit gateway attachment id not found in accepter account ${accepterAccountId} for requester transit gateway attachment id ${requesterTransitGatewayAttachmentId}`,
          );
        }

        attachmentStatus = await getAttachmentState(accepterEc2Client, accepterTransitGatewayAttachmentId);

        await checkAttachmentInvalidStateForAcceptance(accepterEc2Client, accepterTransitGatewayAttachmentId);

        while (attachmentStatus !== 'pendingAcceptance' && attachmentStatus !== 'available') {
          await delay(1000);
          console.log(
            `Waiting for transit gateway attachment id ${accepterTransitGatewayAttachmentId} to be in pendingAcceptance or available state in accepter account ${accepterAccountId}, current state is : ${attachmentStatus}`,
          );
          attachmentStatus = await getAttachmentState(accepterEc2Client, accepterTransitGatewayAttachmentId);
          await checkAttachmentInvalidStateForAcceptance(accepterEc2Client, accepterTransitGatewayAttachmentId);
        }

        if (attachmentStatus !== 'available') {
          console.log(
            `Attachment state is : ${attachmentStatus} in accepter account ${accepterAccountId} for transit gateway attachment id ${accepterTransitGatewayAttachmentId}, proceeding with acceptTransitGatewayPeeringAttachment operation`,
          );

          const response = await throttlingBackOff(() =>
            accepterEc2Client
              .acceptTransitGatewayPeeringAttachment({ TransitGatewayAttachmentId: accepterTransitGatewayAttachmentId })
              .promise(),
          );

          console.log(
            `Transit gateway attachment id ${accepterTransitGatewayAttachmentId} acceptTransitGatewayPeeringAttachment operation completed in accepter account ${accepterAccountId}, waiting to be in available state, current state is ${response.TransitGatewayPeeringAttachment?.State}`,
          );

          attachmentStatus = await getAttachmentState(accepterEc2Client, accepterTransitGatewayAttachmentId);
          while (attachmentStatus !== 'available') {
            await delay(60000);
            console.log(
              `Waiting for attachment id ${accepterTransitGatewayAttachmentId} in accepter account ${accepterAccountId} to be in available state, current state is : ${attachmentStatus}`,
            );
            attachmentStatus = await getAttachmentState(accepterEc2Client, accepterTransitGatewayAttachmentId);
          }
        } else {
          console.log(
            `Attachment state is : ${attachmentStatus} in accepter account ${accepterAccountId} for transit gateway attachment id ${accepterTransitGatewayAttachmentId}, acceptTransitGatewayPeeringAttachment operation not needed`,
          );
        }
      }
      console.log(`Starting route table association`);

      // Associate requester attachment to route table
      await associateTransitGatewayRouteTable(
        requesterAccountId,
        requesterTransitGatewayAttachmentId,
        requesterTransitGatewayRouteTableId,
        requesterEc2Client,
      );

      if (accepterTransitGatewayAttachmentId) {
        // Associate accepter attachment to route table
        await associateTransitGatewayRouteTable(
          accepterAccountId,
          accepterTransitGatewayAttachmentId,
          accepterTransitGatewayRouteTableId,
          accepterEc2Client,
        );

        // Create tags in accepter attachment
        await createAccepterAttachmentTags(accepterEc2Client, accepterTransitGatewayAttachmentId, peeringTags);
      }

      return {
        PhysicalResourceId: requesterTransitGatewayAttachmentId,
        Status: 'SUCCESS',
      };

    case 'Delete':
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}

/**
 * Function to associate transit gateway route table
 * @param account
 * @param transitGatewayAttachmentId
 * @param transitGatewayRouteTableId
 * @param ec2Client
 */
async function associateTransitGatewayRouteTable(
  account: string,
  transitGatewayAttachmentId: string,
  transitGatewayRouteTableId: string,
  ec2Client: AWS.EC2,
): Promise<void> {
  console.log(
    `In Account ${account}: Start of associateTransitGatewayRouteTable, for transitGateway attachment ${transitGatewayAttachmentId} with route table ${transitGatewayRouteTableId}`,
  );

  // Get attachment association information
  const attachmentAssociation = await getAttachmentAssociation(ec2Client, transitGatewayAttachmentId, account);

  //
  // When NO association found OR given route table disassociated
  // Then associate the route table
  if (
    !attachmentAssociation ||
    (attachmentAssociation.State === 'disassociated' &&
      attachmentAssociation.TransitGatewayRouteTableId === transitGatewayRouteTableId)
  ) {
    console.log(
      `Start: Association of transitGateway attachment ${transitGatewayAttachmentId} to route table ${transitGatewayRouteTableId}, in account ${account}`,
    );
    const response = await throttlingBackOff(() =>
      ec2Client
        .associateTransitGatewayRouteTable({
          TransitGatewayAttachmentId: transitGatewayAttachmentId,
          TransitGatewayRouteTableId: transitGatewayRouteTableId,
        })
        .promise(),
    );
    console.log(
      `Completed: Association of transitGateway attachment ${transitGatewayAttachmentId} to route table ${transitGatewayRouteTableId}, in account ${account}, association state is ${response.Association?.State}`,
    );
  } else if (
    attachmentAssociation.State === 'associated' &&
    attachmentAssociation.TransitGatewayRouteTableId !== transitGatewayRouteTableId
  ) {
    // When association found but update requested to change associated route table
    // Disassociate current route table and associate to given route table

    console.log(
      `Route table ${attachmentAssociation.TransitGatewayRouteTableId} already associated with transit gateway attachment ${transitGatewayAttachmentId} in account ${account}`,
    );
    console.log(
      `Start: disassociation of route table ${attachmentAssociation.TransitGatewayRouteTableId} from transit gateway attachment ${transitGatewayAttachmentId} in account ${account}`,
    );
    const disassociationResponse = await throttlingBackOff(() =>
      ec2Client
        .disassociateTransitGatewayRouteTable({
          TransitGatewayAttachmentId: transitGatewayAttachmentId,
          TransitGatewayRouteTableId: attachmentAssociation.TransitGatewayRouteTableId!,
        })
        .promise(),
    );
    console.log(
      `Completed: disassociation of route table ${attachmentAssociation.TransitGatewayRouteTableId} from transit gateway attachment ${transitGatewayAttachmentId} in account ${account}, association state is ${disassociationResponse.Association?.State}`,
    );

    console.log(
      `Start: Re-association of transitGateway attachment ${transitGatewayAttachmentId} to route table ${transitGatewayRouteTableId}, in account ${account}`,
    );

    let reAssociationStatus = await reAssociateRouteTable(
      ec2Client,
      transitGatewayAttachmentId,
      transitGatewayRouteTableId,
      account,
    );

    // Disassociation takes time so sleep and try to re-associate
    while (!reAssociationStatus) {
      console.log(
        'Re-association failed because disassociation still in-progress, sleeping 1 minute before starting again',
      );
      await delay(60000);
      reAssociationStatus = await reAssociateRouteTable(
        ec2Client,
        transitGatewayAttachmentId,
        transitGatewayRouteTableId,
        account,
      );
    }
  }
}

/**
 * Function to re-associate route table
 * @param ec2Client
 * @param transitGatewayAttachmentId
 * @param transitGatewayRouteTableId
 * @param account
 * @returns
 */
async function reAssociateRouteTable(
  ec2Client: AWS.EC2,
  transitGatewayAttachmentId: string,
  transitGatewayRouteTableId: string,
  account: string,
): Promise<boolean> {
  try {
    const associateResponse = await throttlingBackOff(() =>
      ec2Client
        .associateTransitGatewayRouteTable({
          TransitGatewayAttachmentId: transitGatewayAttachmentId,
          TransitGatewayRouteTableId: transitGatewayRouteTableId,
        })
        .promise(),
    );
    console.log(
      `Completed: Re-association of transitGateway attachment ${transitGatewayAttachmentId} to route table ${transitGatewayRouteTableId}, in account ${account}, association state is ${associateResponse.Association?.State}`,
    );
    return true;
  } catch (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    e: any
  ) {
    if (
      // SDKv2 Error Structure
      e.code === 'Resource.AlreadyAssociated' ||
      // SDKv3 Error Structure
      e.name === 'Resource.AlreadyAssociated'
    ) {
      return false;
    }
    throw new Error(e);
  }
}

/**
 * Function to get accepter transit gateway attachment id
 * @param ec2Client
 * @param accepterTransitGatewayId
 * @param accepterAccountId
 * @param requesterTransitGatewayId
 * @param requesterTransitGatewayAttachmentId
 * @returns
 */
async function getAccepterTransitGatewayAttachmentID(
  ec2Client: AWS.EC2,
  accepterTransitGatewayId: string,
  accepterAccountId: string,
  requesterTransitGatewayId: string,
  requesterTransitGatewayAttachmentId: string,
): Promise<string | undefined> {
  const response = await throttlingBackOff(() =>
    ec2Client
      .describeTransitGatewayAttachments({
        Filters: [
          { Name: 'resource-type', Values: ['peering'] },
          { Name: 'transit-gateway-id', Values: [accepterTransitGatewayId] },
          { Name: 'resource-id', Values: [requesterTransitGatewayId] },
          {
            Name: 'state',
            Values: ['pendingAcceptance', 'pending', 'initiatingRequest', 'initiating', 'modifying', 'available'],
          },
        ],
      })
      .promise(),
  );
  if (response.TransitGatewayAttachments) {
    if (response.TransitGatewayAttachments.length === 1) {
      return response.TransitGatewayAttachments[0].TransitGatewayAttachmentId;
    }

    if (response.TransitGatewayAttachments.length === 0) {
      throw new Error(
        `Transit gateway attachment id not found in accepter account ${accepterAccountId} for requester transit gateway attachment id ${requesterTransitGatewayAttachmentId}`,
      );
    }

    if (response.TransitGatewayAttachments.length > 1) {
      const id: string[] = [];
      for (const item of response.TransitGatewayAttachments) {
        id.push(item.TransitGatewayAttachmentId!);
      }
      throw new Error(
        `Multiple transit gateway attachment ids ${id} found in accepter account ${accepterAccountId} for requester transit gateway attachment id ${requesterTransitGatewayAttachmentId}`,
      );
    }
  }
  return undefined;
}

/**
 * Function to get attachment status
 * @param transitGatewayAttachmentId
 * @param ec2Client
 * @returns
 */
async function getAttachmentState(
  ec2Client: AWS.EC2,
  transitGatewayAttachmentId: string,
): Promise<AWS.EC2.TransitGatewayAttachmentState | undefined> {
  const response = await throttlingBackOff(() =>
    ec2Client
      .describeTransitGatewayPeeringAttachments({
        Filters: [{ Name: 'transit-gateway-attachment-id', Values: [transitGatewayAttachmentId] }],
      })
      .promise(),
  );

  if (response.TransitGatewayPeeringAttachments?.length === 1) {
    console.log(
      `Transit gateway attachment ${transitGatewayAttachmentId} state is ${response.TransitGatewayPeeringAttachments[0].State}`,
    );
    return response.TransitGatewayPeeringAttachments[0].State;
  }

  return undefined;
}

/**
 * Function to check for Invalid attachment state for acceptance
 * @param ec2Client
 * @param transitGatewayAttachmentId
 */
async function checkAttachmentInvalidStateForAcceptance(
  ec2Client: AWS.EC2,
  transitGatewayAttachmentId: string,
): Promise<void> {
  const response = await throttlingBackOff(() =>
    ec2Client
      .describeTransitGatewayPeeringAttachments({
        Filters: [{ Name: 'transit-gateway-attachment-id', Values: [transitGatewayAttachmentId] }],
      })
      .promise(),
  );

  if (response.TransitGatewayPeeringAttachments?.length === 1) {
    if (
      response.TransitGatewayPeeringAttachments[0].State === 'failed' ||
      response.TransitGatewayPeeringAttachments[0].State === 'failing' ||
      response.TransitGatewayPeeringAttachments[0].State === 'deleted' ||
      response.TransitGatewayPeeringAttachments[0].State === 'deleting' ||
      response.TransitGatewayPeeringAttachments[0].State === 'rejected' ||
      response.TransitGatewayPeeringAttachments[0].State === 'rejecting' ||
      response.TransitGatewayPeeringAttachments[0].State === 'rollingBack'
    ) {
      throw new Error(
        `Transit gateway attachment ${transitGatewayAttachmentId} in ${response.TransitGatewayPeeringAttachments[0].State}, acceptance of attachment can not be performed !!!`,
      );
    }
  }
}

/**
 * Function to get transit gateway attachment associated route table id
 * @param ec2Client
 * @param transitGatewayAttachmentId
 * @param account
 * @returns
 */
async function getAttachmentAssociation(
  ec2Client: AWS.EC2,
  transitGatewayAttachmentId: string,
  account: string,
): Promise<AWS.EC2.TransitGatewayAttachmentAssociation | undefined> {
  const response = await throttlingBackOff(() =>
    ec2Client
      .describeTransitGatewayAttachments({
        Filters: [
          { Name: 'resource-type', Values: ['peering'] },
          { Name: 'transit-gateway-attachment-id', Values: [transitGatewayAttachmentId] },
        ],
      })
      .promise(),
  );

  if (response.TransitGatewayAttachments) {
    if (response.TransitGatewayAttachments.length === 1) {
      return response.TransitGatewayAttachments[0].Association;
    } else if (response.TransitGatewayAttachments.length === 0) {
      return undefined;
    } else {
      throw new Error(
        `Multiple transit gateway attachments ${transitGatewayAttachmentId} found in account ${account} !!!`,
      );
    }
  } else {
    throw new Error(`Transit gateway attachment ${transitGatewayAttachmentId} not found in account ${account} !!!`);
  }
}

/**
 * Function to create tags for accepter transit gateway peering attachment
 * @param ec2Client
 * @param accepterTransitGatewayAttachmentId
 */
async function createAccepterAttachmentTags(
  ec2Client: AWS.EC2,
  accepterTransitGatewayAttachmentId: string,
  peeringTags: { Key: string; Value: string }[] | undefined,
): Promise<void> {
  const tags: { Key: string; Value: string }[] = [];

  if (peeringTags) {
    for (const peeringTag of peeringTags ?? []) {
      tags.push(peeringTag as { Key: string; Value: string });
    }

    if (tags.length > 0) {
      await throttlingBackOff(() =>
        ec2Client
          .createTags({
            Resources: [accepterTransitGatewayAttachmentId],
            Tags: tags,
          })
          .promise(),
      );
    }
  }
}

/**
 * Function to create accepter EC2 client
 * @param accepterRegion
 * @param accepterRoleArn
 * @returns
 */
async function getAccepterEc2Client(
  accepterRegion: string,
  accepterRoleArn: string,
  solutionId?: string,
): Promise<AWS.EC2> {
  const stsClient = new AWS.STS({ customUserAgent: solutionId, region: accepterRegion });

  const assumeRoleResponse = await throttlingBackOff(() =>
    stsClient
      .assumeRole({
        RoleArn: accepterRoleArn,
        RoleSessionName: 'AccepterTransitGatewayAttachmentSession',
      })
      .promise(),
  );

  return new AWS.EC2({
    credentials: {
      accessKeyId: assumeRoleResponse.Credentials?.AccessKeyId ?? '',
      secretAccessKey: assumeRoleResponse.Credentials?.SecretAccessKey ?? '',
      sessionToken: assumeRoleResponse.Credentials?.SessionToken,
    },
    region: accepterRegion,
  });
}

/**
 * Function to sleep process
 * @param ms
 * @returns
 */
function delay(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
