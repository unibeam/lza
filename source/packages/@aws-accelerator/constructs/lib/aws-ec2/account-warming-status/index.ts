/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * aws-ec2-pre-warm-account-status - lambda handler
 *
 * @param event
 * @returns
 */

import { SSMClient, PutParameterCommand } from '@aws-sdk/client-ssm';
import {
  EC2Client,
  DescribeInstancesCommand,
  DescribeVpcsCommand,
  DescribeSubnetsCommand,
  DeleteVpcCommand,
  DeleteSubnetCommand,
  TerminateInstancesCommand,
} from '@aws-sdk/client-ec2';
import { AdaptiveRetryStrategy } from '@aws-sdk/util-retry';
import { delay } from '@aws-accelerator/utils/lib/throttle';

const solutionId = process.env['SOLUTION_ID'] ?? '';
const retryStrategy = new AdaptiveRetryStrategy(() => Promise.resolve(5));
const ec2Client = new EC2Client({ customUserAgent: solutionId, retryStrategy });
const ssmClient = new SSMClient({ customUserAgent: solutionId, retryStrategy });

type InstanceDetails = {
  instanceId: string | undefined;
  minutesSinceLaunch: number | undefined;
};

//eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(event: any): Promise<
  | {
      IsComplete: boolean;
    }
  | undefined
> {
  console.log(event);
  const ssmPrefix: string = event.ResourceProperties['ssmPrefix'];
  const instanceDetails = await getInstanceDetails();

  // if no instance was found, it has been deleted return
  if (instanceDetails.instanceId === undefined) {
    return { IsComplete: true };
  }

  if (instanceDetails.minutesSinceLaunch! > 15) {
    console.log('Account warming time reached');
    await terminateInstance(instanceDetails.instanceId);
    await deleteVpc();
    await setSSMParameter(ssmPrefix, 'true');
    return {
      IsComplete: true,
    };
  }
  console.log('Waiting for warmup to complete');
  return { IsComplete: false };
}

async function getInstanceDetails(): Promise<InstanceDetails> {
  console.log('Getting Instance Id');
  const instanceDetails: InstanceDetails = { instanceId: undefined, minutesSinceLaunch: undefined };
  const ec2Instances = await ec2Client.send(
    new DescribeInstancesCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (ec2Instances.Reservations && ec2Instances.Reservations.length < 1) {
    return instanceDetails;
  }

  if (ec2Instances.Reservations![0].Instances?.length ?? 0 > 0) {
    instanceDetails.instanceId = ec2Instances.Reservations![0].Instances![0].InstanceId;
    const launchTime = ec2Instances.Reservations![0].Instances![0].LaunchTime;

    const currentTime = new Date();
    const msElapsed = currentTime.getTime() - launchTime!.getTime();
    instanceDetails.minutesSinceLaunch = msElapsed / 1000 / 60;
  }

  return instanceDetails;
}

async function deleteVpc() {
  console.log('Deleting VPC');
  const vpcId = await getVpcId();

  if (vpcId) {
    console.log('Deleting subnets');
    const subnets = await ec2Client.send(
      new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }),
    );
    for (const subnet of subnets.Subnets) {
      await ec2Client.send(new DeleteSubnetCommand({ SubnetId: subnet.SubnetId }));
    }
    console.log(`Deleting VPC with id: ${vpcId}`);
    await ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId }));
  }
}

async function getVpcId(): Promise<string | undefined> {
  console.log('Getting VPC Id');
  const vpcs = await ec2Client.send(
    new DescribeVpcsCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (vpcs.Vpcs?.length ?? 0 > 0) {
    return vpcs.Vpcs![0].VpcId;
  }
  return undefined;
}

async function setSSMParameter(ssmPrefix: string, parameterValue: string) {
  console.log('Updating SSM Parameter');
  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: `${ssmPrefix}/account/pre-warmed`,
        Value: parameterValue,
        Overwrite: true,
      }),
    );
  } catch (e) {
    console.log(e);
  }
}

async function terminateInstance(instanceId: string): Promise<void> {
  console.log(`Terminating EC2 instance: ${instanceId}`);
  await ec2Client.send(new TerminateInstancesCommand({ InstanceIds: [instanceId] }));
  await waitForTermination(instanceId);
}

async function waitForTermination(instanceId: string): Promise<void> {
  console.log(`Waiting for termination of instanceId: ${instanceId}`);
  let ec2Terminated = false;
  while (!ec2Terminated) {
    const statusResponse = await ec2Client.send(new DescribeInstancesCommand({ InstanceIds: [instanceId] }));
    if (
      statusResponse.Reservations[0].Instances[0].State?.Name !== 'terminated' ||
      statusResponse.Reservations[0].Instances[0].State?.Code !== 48
    ) {
      delay(15000);
      continue;
    }
    ec2Terminated = true;
  }
}
