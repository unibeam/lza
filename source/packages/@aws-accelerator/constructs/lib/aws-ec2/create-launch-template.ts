/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';
import { NetworkInterfaceItemConfig } from '@aws-accelerator/config';

export interface ILaunchTemplateResource extends cdk.IResource {
  /**
   * The version number.
   */
  readonly version: string;
  /**
   * The ID of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateId: string;
  /**
   * The name of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateName: string;
}

export type EbsProperty = {
  deleteOnTermination?: boolean;
  encrypted?: boolean;
  iops?: number;
  kmsKeyId?: string;
  snapshotId?: string;
  throughput?: number;
  volumeSize?: number;
  volumeType?: string;
};

export type BlockDeviceMappingItem = {
  deviceName: string;
  ebs?: EbsProperty;
};

export interface LaunchTemplateProps {
  /*
   * Either the path to the user data file or a base-64 encoded userdata string
   */
  readonly userData?: string;
  /*
   * VpcName
   */
  readonly vpc: string;
  /*
   * Name of Launch Template
   */
  readonly name: string;
  /*
   * Name of Application
   */
  readonly appName: string;
  readonly blockDeviceMappings?: BlockDeviceMappingItem[];
  readonly securityGroups?: string[];
  readonly networkInterfaces?: NetworkInterfaceItemConfig[];
  readonly instanceType: string;
  readonly keyPair?: string;
  readonly iamInstanceProfile?: string;
  readonly imageId: string;
  readonly enforceImdsv2?: boolean;
}

export class LaunchTemplate extends cdk.Resource implements ILaunchTemplateResource {
  public readonly version: string;
  public readonly launchTemplateId: string;
  public readonly launchTemplateName: string;

  constructor(scope: Construct, id: string, props: LaunchTemplateProps) {
    super(scope, id);

    let metadataOptions: cdk.aws_ec2.CfnLaunchTemplate.MetadataOptionsProperty;

    if (props.enforceImdsv2 === false) {
      metadataOptions = { httpTokens: 'optional' };
    } else {
      metadataOptions = { httpTokens: 'required' };
    }

    let networkInterfaceProps: cdk.aws_ec2.CfnLaunchTemplate.NetworkInterfaceProperty[] | undefined = undefined;
    if (props.networkInterfaces) {
      networkInterfaceProps = this.setNetworkInterfaceProps(props.networkInterfaces);
    }

    const launchTemplate = new cdk.aws_ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        securityGroupIds: props.securityGroups ?? undefined,
        networkInterfaces: networkInterfaceProps ?? undefined,
        instanceType: props.instanceType,
        keyName: props.keyPair ?? undefined,
        imageId: props.imageId,
        iamInstanceProfile: { name: props.iamInstanceProfile ?? undefined },
        userData: this.processUserData(props.userData),
        metadataOptions,
      },
      launchTemplateName: props.name,
    });

    this.launchTemplateId = launchTemplate.ref;
    this.launchTemplateName = launchTemplate.launchTemplateName!;
    this.version = launchTemplate.attrLatestVersionNumber;
  }

  private setNetworkInterfaceProps(
    interfaces: NetworkInterfaceItemConfig[],
  ): cdk.aws_ec2.CfnLaunchTemplate.NetworkInterfaceProperty[] {
    const networkInterfaceProps: cdk.aws_ec2.CfnLaunchTemplate.NetworkInterfaceProperty[] = [];
    for (const networkInterface of interfaces) {
      networkInterfaceProps.push({
        associateCarrierIpAddress: networkInterface.associateCarrierIpAddress,
        associatePublicIpAddress: networkInterface.associatePublicIpAddress,
        deleteOnTermination: networkInterface.deleteOnTermination,
        description: networkInterface.description,
        deviceIndex: networkInterface.deviceIndex,
        groups: networkInterface.groups,
        interfaceType: networkInterface.interfaceType,
        networkCardIndex: networkInterface.networkCardIndex,
        networkInterfaceId: networkInterface.networkInterfaceId,
        privateIpAddress: networkInterface.privateIpAddress,
        privateIpAddresses: networkInterface.privateIpAddresses,
        secondaryPrivateIpAddressCount: networkInterface.secondaryPrivateIpAddressCount,
        subnetId: networkInterface.subnetId,
      });
    }
    return networkInterfaceProps;
  }

  /**
   * Process user data file or string
   * @param userDataStr string | undefined
   * @returns string | undefined
   */
  private processUserData(userDataStr?: string): string | undefined {
    if (!userDataStr) {
      return;
    }
    //
    // If string is a token, return it. It has already been processed upstream
    if (cdk.Token.isUnresolved(userDataStr)) {
      return userDataStr;
    } else {
      //
      // Else, userDataStr is a file path, return it in base-64 encoding.
      return cdk.Fn.base64(fs.readFileSync(userDataStr, 'utf8'));
    }
  }
}
