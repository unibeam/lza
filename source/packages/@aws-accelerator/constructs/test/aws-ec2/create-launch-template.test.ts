/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { LaunchTemplate } from '../../lib/aws-ec2/create-launch-template';
import { snapShotTest } from '../snapshot-test';
import * as path from 'path';
import { describe } from '@jest/globals';

const testNamePrefix = 'Construct(LaunchTemplate): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new LaunchTemplate(stack, 'Test', {
  name: 'Test',
  appName: 'appA',
  userData: path.join(__dirname, 'launchTemplateFiles/testUserData.sh'),
  vpc: 'test',
  instanceType: 't3.micro',
  imageId: 'ami-1234',
  securityGroups: ['sg-1234'],
  blockDeviceMappings: [
    {
      deviceName: 'test',
      ebs: { deleteOnTermination: true, encrypted: true, iops: 300, kmsKeyId: 'test', volumeSize: 10 },
    },
  ],
});

/**
 * GWLB construct test
 */
describe('LaunchTemplate', () => {
  snapShotTest(testNamePrefix, stack);
});
