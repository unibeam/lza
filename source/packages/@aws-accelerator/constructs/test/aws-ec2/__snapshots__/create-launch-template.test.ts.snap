// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchTemplate Construct(LaunchTemplate):  Snapshot Test 1`] = `
{
  "Resources": {
    "TestLaunchTemplateAC264C67": {
      "Properties": {
        "LaunchTemplateData": {
          "BlockDeviceMappings": [
            {
              "DeviceName": "test",
              "Ebs": {
                "DeleteOnTermination": true,
                "Encrypted": true,
                "Iops": 300,
                "KmsKeyId": "test",
                "VolumeSize": 10,
              },
            },
          ],
          "IamInstanceProfile": {},
          "ImageId": "ami-1234",
          "InstanceType": "t3.micro",
          "MetadataOptions": {
            "HttpTokens": "required",
          },
          "SecurityGroupIds": [
            "sg-1234",
          ],
          "UserData": {
            "Fn::Base64": "#!/bin/bash
yum update -y
yum install -y httpd 
systemctl start httpd
systemctl enable httpd
usermod -a -G apache ec2-user
chown -R ec2-user:apache /var/www
chmod 2775 /var/www
find /var/www -type d -exec chmod 2775 {} \\;
find /var/www -type f -exec chmod 0664 {} \\;",
          },
        },
        "LaunchTemplateName": "Test",
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
  },
}
`;
